From 205df44462530825ded3ffe3e2d6d68cb716287d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?St=C3=A9phane=20Graber?= <stgraber@ubuntu.com>
Date: Thu, 5 Nov 2020 15:17:40 -0500
Subject: [PATCH lxcfs 4/8] swap: Fix usage reporting
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Stéphane Graber <stgraber@ubuntu.com>
---
 src/proc_fuse.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/proc_fuse.c b/src/proc_fuse.c
index 7bc1eb0..0ad21b9 100644
--- a/src/proc_fuse.c
+++ b/src/proc_fuse.c
@@ -407,7 +407,7 @@ static int proc_swaps_read(char *buf, size_t size, off_t offset,
 	if (swtotal > 0) {
 		l = snprintf(d->buf + total_len, d->size - total_len,
 			     "none%*svirtual\t\t%" PRIu64 "\t%" PRIu64 "\t0\n",
-			     36, " ", swtotal, swfree);
+			     36, " ", swtotal, swusage);
 		total_len += l;
 	}
 
-- 
2.20.1

