From 1d3a4a0dd7411d34f10e47c1d850213a63a3581e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?St=C3=A9phane=20Graber?= <stgraber@ubuntu.com>
Date: Thu, 5 Nov 2020 15:46:07 -0500
Subject: [PATCH lxcfs 8/8] swap: Remove now unused variable
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Stéphane Graber <stgraber@ubuntu.com>
---
 src/proc_fuse.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/src/proc_fuse.c b/src/proc_fuse.c
index fc69789..9bf963b 100644
--- a/src/proc_fuse.c
+++ b/src/proc_fuse.c
@@ -325,7 +325,7 @@ static int proc_swaps_read(char *buf, size_t size, off_t offset,
 	bool wants_swap = opts && !opts->swap_off && liblxcfs_can_use_swap();
 	struct file_info *d = INTTYPE_TO_PTR(fi->fh);
 	uint64_t memswlimit = 0, memlimit = 0, memusage = 0, memswusage = 0,
-		 swtotal = 0, swfree = 0, swusage = 0, memswpriority = 1
+		 swtotal = 0, swusage = 0, memswpriority = 1,
 		 hostswtotal = 0, hostswfree = 0;
 	ssize_t total_len = 0;
 	ssize_t l = 0;
@@ -383,8 +383,6 @@ static int proc_swaps_read(char *buf, size_t size, off_t offset,
 					swusage = 0;
 				else
 					swusage = (memswusage - memusage) / 1024;
-				if (swtotal >= swusage)
-					swfree = swtotal - swusage;
 			}
 
 			ret = cgroup_ops->get_memory_swappiness(cgroup_ops, cgroup, &memswpriority_str);
-- 
2.20.1

