From 3f4847aacd7cb56d7859f1562c81c06acf9d41d5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fabian=20Gr=C3=BCnbichler?= <f.gruenbichler@proxmox.com>
Date: Wed, 19 Oct 2016 09:16:36 +0200
Subject: [PATCH] fix Active/Inactive /proc/meminfo

---
 bindings.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/bindings.c b/bindings.c
index 2f78ab5..0d46fb8 100644
--- a/bindings.c
+++ b/bindings.c
@@ -3197,11 +3197,11 @@ static int proc_meminfo_read(char *buf, size_t size, off_t offset,
 		} else if (startswith(line, "SwapCached:")) {
 			snprintf(lbuf, 100, "SwapCached:     %8lu kB\n", 0UL);
 			printme = lbuf;
-		} else if (startswith(line, "Active")) {
+		} else if (startswith(line, "Active:")) {
 			snprintf(lbuf, 100, "Active:         %8lu kB\n",
 					active_anon + active_file);
 			printme = lbuf;
-		} else if (startswith(line, "Inactive")) {
+		} else if (startswith(line, "Inactive:")) {
 			snprintf(lbuf, 100, "Inactive:       %8lu kB\n",
 					inactive_anon + inactive_file);
 			printme = lbuf;
-- 
2.1.4

