From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Wolfgang Bumiller <w.bumiller@proxmox.com>
Date: Thu, 2 Dec 2021 13:47:24 +0100
Subject: [PATCH lxcfs] build with fuse2

this stuff is a PITA

Signed-off-by: Wolfgang Bumiller <w.bumiller@proxmox.com>
---
 configure.ac | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/configure.ac b/configure.ac
index ccea4ec..9260137 100644
--- a/configure.ac
+++ b/configure.ac
@@ -61,13 +61,7 @@ AC_CONFIG_FILES([
 
 AC_CHECK_LIB(pthread, main)
 
-PKG_CHECK_MODULES(FUSE, fuse3, [have_fuse3=yes], [have_fuse3=no])
-if test "x$have_fuse3" = "xyes"; then
-	AC_DEFINE([HAVE_FUSE3], [1], [Use fuse3])
-	AC_DEFINE_UNQUOTED([FUSE_USE_VERSION], [30], [Use fuse version 30])
-else
 	PKG_CHECK_MODULES(FUSE, fuse, [AC_DEFINE_UNQUOTED([FUSE_USE_VERSION], [26], [Use fuse])])
-fi
 
 AC_PATH_PROG(HELP2MAN, help2man, false // No help2man //)
 AM_CONDITIONAL([HAVE_HELP2MAN], [test "x$HELP2MAN" != "xfalse // No help2man //" ])
